Solving issues with large rotational limits

francofusco opened the pull request:
As found out by others as well, there is a small issue in some controllers when a revolute joints has "large" limits, see eg #365 

The problem is that some position controllers call the function `angles::shortest_angular_distance_with_limits`, which assumes that the total allowed range of rotation is lower than `2*PI`. Therefore, if one has as limits such as `[-PI-eps, PI+eps)`, `angles` will assume that the valid range has size `2*eps` instead of `2*(PI+eps)`. To allow dealing with larger rotations, the PR ros/angles#16 introduced the function `shortest_angular_distance_with_large_limits`.

I updated three controllers to use such function in case `upper_limit-lower_limit` is greater than or equal to `2*PI`. I tested it and the oscillations that could be experienced before now disappear.

francofusco committed:
Solving issues with large rotational limits

bmagyar reviewed (changes requested):
The changes look good to me on first read. Could you please add a test that fails without these changes? Could reuse the same code for all controllers.

francofusco commented:
Glad to! However, how should I proceed? The issue can be seen "visually" without much problems, however I am not so sure about what strategy to follow to analyze it in a test - I believe you want a test that can be checked by continuous integration, isn't it? I think the easier way to spot the problem in the code is to check the return value of `angles::shortest_angular_distance_with_limits` (it will likely return `false` at some point), but I am not sure that it will always do so... I will think about it and try to come up with a satisfactory solution!

matthew-reynolds commented on effort_controllers/src/joint_group_position_controller.cpp:
```diff
@@ -138,12 +138,24 @@ namespace effort_controllers
+              joint_urdfs_[i]->limits->lower,
+              joint_urdfs_[i]->limits->upper,
+              error);
+          }
```
Is it necessary to perform this check? Since `ros_control` would have several issues if `lower < upper` is violated, I think it is safe to assume that `lower` will always be less than `upper`. As such, wouldn't it be equivalent to just always use the new `shortest_angular_distance_with_large_limits`?

francofusco replied:
No, I don't think it is necessary - I left it since I was not entirely sure if at any point one could have `lower > upper`. I will remove it!

matthew-reynolds reviewed (commented):
Looks good to me, I agree that tests would be great. Perhaps we could check the sign of the commanded vel/effort for a specific set of limits and current position to see if it is using the correct limits? I don't know what angle `shortest_angular_distance_with_limits` returns if the limits are violated, so maybe this wouldn't work.

bmagyar commented:
Ping

francofusco committed:
Assuming lower limits are smaller than upper limits

francofusco committed:
Merge branch 'melodic-devel' of https://github.com/ros-controls/ros_controllers into large_angular_limits

francofusco commented:
Sorry for being inactive for such a long time.

I removed the check `lower < upper`, I will now work on a test to make sure things work as intended.

francofusco committed:
Added test for position controller

francofusco committed:
test_depend on robot_state_publisher

francofusco commented:
I added a simple test for `effort_controllers::JointPositionController`: it involves a simple robot with a single revolute joint with limits `[-6.3, 6.3]`. I created a supersimple `RobotHW` to fake a real robot, and the test simply tries to move the single joint to the positions `1.0`, `-1.0` and `0.0` in succession. Without the fix, the joint oscillates around the initial position and thus never reaches the targets - teh tset should fail due to the 50s timeout set in `[effort_position_controller.test](effort_controllers/test/effort_position_controller.test)`. With the fix, the position is reached and the test should run just fine. To "see" the test, you can launch the test manually:
```
roslaunch effort_controllers simple_bot_hw.launch
rosrun effort_controllers effort_position_controller_test
```
If you start RViz, set the fixed frame as `world` and add a `RobotModel` plugin you should be able to see the robot (nothing more than a rotating beam). Recompile with and without the fix to see the difference. 

Is this good enough? The test is quite simple, but I could not think of any better solution... If you think it is ok, I can then add a similar test also for `velocity_controllers/JointPositionController`.

Note: the test is failing during the CI due to the timeout. Since on my PC it was working well, I think it might be due to a lower processing power of the hosting virtual machine (?). I will try to solve the issue by increasing both the control period (it is currently 1ms) and the timeout. Sorry if this adds unnecessary commits...

francofusco committed:
more "permissive" tests

francofusco committed:
Use sim time in test

francofusco committed:
provide error message in test

francofusco committed:
Start controllers before test

francofusco committed:
limit commanded effort

francofusco commented:
I was finally able to make the tests succeed. Again, I apologize for the long commit list... I can rebase and squash all test-related commits into one, let me know!

francofusco commented:
Thanks for the approval! Before merging, do you want me to rebase the commit history and/or add the same test also for `velocity_controllers/JointPositionController`?

matthew-reynolds commented on effort_controllers/test/simple_bot_goto_position.cpp:
```diff
@@ -0,0 +1,70 @@
+#include <ros/ros.h>
```
Missing licence comment

francofusco replied:
I added the license and I also updated the one of `simple_bot_hw.cpp`. They are now a copy of the one from `joint_position_controller.cpp` with my name added. I am never 100% sure about licenses, so feel free to correct me if something is wrong with this.

matthew-reynolds replied:
I'm never sure either :)

@bmagyar look good?

bmagyar replied:
TBH I wouldn't be too fussed about putting license notes in every file, especially since the package.xml states BSD3. If you'd like though, that's ok but don't attribute new files to Willow Garage or HiDof, if it's just you, put yourself.

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+// ROS
+#include <chrono>
+#include <thread>
+#include <ros/ros.h>
```
Nitpick: `chrono` and `thread` aren't ROS packages.

francofusco replied:
Absolutely right, I moved them up, adding the `#include <cmath>` there as well.

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+  void write() {
+    // store current command
+    eff_ = cmd_;
+    if(std::fabs(eff_) > max_eff_) {
```
Let's `#include <cmath>` for this `fabs` and the `pow` above.

francofusco replied:
See [comment on `#include <chrono>` and `#include <thread>`](https://github.com/ros-controls/ros_controllers/pull/466#discussion_r406656321).

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+    eff_ = cmd_;
+    if(std::fabs(eff_) > max_eff_) {
+      ROS_WARN("Large effort command received: %f; cutting down to % f", eff_, max_eff_);
+      eff_ = eff_ > 0 ? max_eff_ : -max_eff_;
```
Nitpick:
```suggestion
      eff_ = eff_ > 0.0 ? max_eff_ : -max_eff_;
```

francofusco replied:
I updated all doubles to be in the format `XX.0`.

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+  ros::AsyncSpinner spinner(2);
+  spinner.start();
+
+  ros::Rate rate(1./bot.period);
```
Nitpick: I don't think we use `1.` anywhere in this codebase, do you mind switching to `1.0`? Same above at line 65.

francofusco replied:
Sorry, taken care of - see [other comment](https://github.com/ros-controls/ros_controllers/pull/466#discussion_r406657023)

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+    double elapsed = std::chrono::duration_cast<std::chrono::duration<double> >((end - begin)).count();
+    if(bot.period > elapsed)
+      std::this_thread::sleep_for(std::chrono::duration<double>(bot.period - elapsed));
+    rosgraph_msgs::Clock clock;
+    clock.clock = ros::Time(sim_time);
+    clock_publisher.publish(clock);
+    sim_time += bot.period;
```
Just curious, why are you managing the sim time yourself rather than just using the ROS clock?

matthew-reynolds replied:
Oh, I see you're just matching what the other tests do. Must be a good reason for it that I just haven't thought of :)

francofusco replied:
To be honest, I decided to put it because in the first versions of the test I was getting no errors on my PC but checks were failing during CI. I imagined that one issue was related to the virtual machine: perhaps it had limited processing power and thus I wanted to "slow down" things accordingly at least in the RobotHW. I saw that other packages were using the simulated clock and I thus used the same procedure. Now that I found I forgot to initialize the `cmd_` variable, I think that was the main source of issues 😅 But I think it might still be useful, in case the VM slows down...

matthew-reynolds commented on effort_controllers/CMakeLists.txt:
```diff
@@ -57,3 +57,18 @@ install(TARGETS ${PROJECT_NAME}
+    test/effort_position_controller.test
+    test/simple_bot_goto_position.cpp
+  )
+  target_link_libraries(effort_position_controller_test ${catkin_LIBRARIES})
```
Careful here, when you `find_package(catkin REQUIRED COMPONENTS ...)` it wipes out `catkin_INCLUDE_DIRS` and `catkin_LIBRARIES` so we lose everything that was imported by the first `find_package` above. We're also missing a few packages.

You could either save a copy of those vars before you `find_package`, or list each package individually like this:

```suggestion
  find_package(rostest REQUIRED)
  
  find_package(controller_manager REQUIRED)
  find_package(hardware_interface REQUIRED)
  find_package(rosgraph_msgs REQUIRED)
  find_package(sensor_msgs REQUIRED)
  find_package(std_msgs REQUIRED)

  include_directories(
    ${controller_manager_INCLUDE_DIRS}
    ${hardware_interface_INCLUDE_DIRS}
    ${rosgraph_msgs_INCLUDE_DIRS}
    ${sensor_msgs_INCLUDE_DIRS)
    ${std_msgs_INCLUDE_DIRS}
  )

  add_executable(simple_bot_hw test/simple_bot_hw.cpp)
  target_link_libraries(simple_bot_hw
    ${catkin_LIBRARIES}
    ${controller_manager_LIBRARIES}
    ${hardware_interface_LIBRARIES}
  )
  add_dependencies(tests simple_bot_hw)

  add_rostest_gtest(effort_position_controller_test
    test/effort_position_controller.test
    test/simple_bot_goto_position.cpp
  )
  target_link_libraries(effort_position_controller_test ${catkin_LIBRARIES})
```

With either approach, please add the missing `hardware_interface`, `sensor_msgs`, and `std_msgs` dependencies, and also pull `find_package(rostest REQUIRED)` out to be its own line, at the beginning of the `if(CATKIN_ENABLE_TESTING)` block.

francofusco replied:
Thanks for pointing this out, I was actually not aware of that! I chose the second approach, so every package is found separately wih its own `find_package`. I sorted the entries alphabetically - while leaving `rostest` on top of everything.

matthew-reynolds replied:
Yep, it's on my TODO list to fix up the other packages that have `find_package(catkin ...` twice as well. Seems like a common problem, it's a shame `find_package(catkin ...` doesn't have an option to append to the lists rather than overwrite. Thanks for fixing it!

matthew-reynolds commented on effort_controllers/package.xml:
```diff
@@ -17,14 +17,20 @@
+  <test_depend>rosgraph_msgs</test_depend>
+  <test_depend>controller_manager</test_depend>
+  <test_depend>xacro</test_depend>
+  <test_depend>robot_state_publisher</test_depend>
```
Please alphabetize and add additional depends as described above.

matthew-reynolds commented on effort_controllers/test/simple_bot_goto_position.cpp:
```diff
@@ -0,0 +1,70 @@
+  ros::NodeHandle nh;
+  double position;
+
+  int val;
```
This looks unused, let's remove it.

francofusco replied:
Totally right.

matthew-reynolds commented on effort_controllers/test/simple_bot_goto_position.cpp:
```diff
@@ -0,0 +1,70 @@
+TEST_F(SimpleBotFixture, TestPosition) {
+  ros::Duration(1.0).sleep();
+  std::string err;
+  ASSERT_TRUE(waitToReachTarget(1.0, err)) << err;
+  ASSERT_TRUE(waitToReachTarget(-1.0, err)) << err;
+  ASSERT_TRUE(waitToReachTarget(0.0, err)) << err;
```
I think these should be `EXPECT`, not `ASSERT`.

francofusco replied:
As far as I understand, the only difference between `ASSERT` and `EXPECT` is that the former will stop the test right away, while the latter will record the failure but let the test finish. As there is no cleanup needed in the test, if one of the three `waitToReachTarget` calls fail, we can already exit. This is why I used `ASSERT` rather than `EXPECT`, but if there is any reason to change it, I will!

matthew-reynolds replied:
Yep, my only thinking was that `EXPECT` might help if there's any flakiness. If 0.1% of the time a test fails, if we use `EXPECT` we can see "Oh, 2/3 passed, probably just need to re-run the test" rather than thinking the changes broke the test entirely.

francofusco replied:
I get it, makes sense! I will change them accordingly

matthew-reynolds commented on effort_controllers/test/simple_bot_hw.cpp:
```diff
@@ -0,0 +1,120 @@
+  : pos_(0)
+  , vel_(0)
+  , eff_(0)
+  , max_eff_(10)
```
Nitpick: Can we add `.0` to all of these?

francofusco replied:
[Done](https://github.com/ros-controls/ros_controllers/pull/466#discussion_r406657023).

matthew-reynolds reviewed (changes requested):
This looks great, thanks! I think we just need to touch up the CMakeLists & `package.xml` to get the dependencies sorted out, and then I have a handful of small requests.

Can you confirm that these tests fail with the old `angles::shortest_angular_distance_with_limits`?

bmagyar reviewed (approved):
I'm happy with these, can merge as soon as @matthew-reynolds is happy too :)

francofusco committed:
update test according to review

francofusco committed:
init cmd_ in SimpleBot

francofusco commented:
@matthew-reynolds I addressed your comments, let me know if anything needs to be updated :smiley: 

> Can you confirm that these tests fail with the old `angles::shortest_angular_distance_with_limits`?

Yes, if I change the `angles` call to the old one, the joint does not rotate (actually, it keeps oscillating around zero).

*Update*: I uploaded two videos, showing the difference between the [proposed changes](https://youtu.be/Z9KjIDEziaM) and the [old behavior](https://youtu.be/Wdn7o0pLc80).

francofusco commented:
Just found a stupid mistake in the `simple_bot_hw.cpp`: the variable `cmd_` was not initialized in the constructor, I believe it was causing troubles in the tests...

matthew-reynolds reviewed (approved):
Awesome, thanks for the videos, this is great.

I'm good as soon as @bmagyar gives the license a peek, and once we swap to `EXPECT`.

francofusco committed:
EXPECT_TRUE instead of ASSERT_TRUE + license

francofusco commented:
I updated the code according to the last comments. I think all points from your reviews have been addressed, but let me know if I forgot something instead!

bmagyar commented:
Thanks very much for sticking with us through this review! Merging...
